### PR TITLE
Dev watchers

### PR DESCRIPTION
**Pull request by printf-twinkle**
### Functionality:
watchers in task details page

### Solution:
1) task creator, assignee and reporter will automatically be added as watchers to the particular task.
2) in case all three are the same or duplicate data is removed.
3) if anybody else wants to keep a track of the task updates, the person can click on the eye button in watchers section so will get all email updates.

### Risk level:
- [ ] high 
- [x] medium
- [ ] low

### How to test:
create a task or update a task

### CODE REVIEW

File: functions/model/tasks/tark/addComment.js [hunk 1]
```diff
@@ -62,41 +64,49 @@ exports.addComment = function(request, response) {
...
 
             addActivity("COMMENT", logWorkComment, taskId, date, time, orgDomain, uid);
+            watchers.forEach(element => {
+                sendMail(element, subjectMessage, htmlMessage);
+                console.log("mail sent!!!!!!!")
+            });
```
Comment (viv9k) [inline functions/model/tasks/tark/addComment.js:91]: No need for these, changes we will take care of it in mailer.

File: functions/model/tasks/tark/deleteTask.js [hunk 2]
```diff
@@ -119,7 +119,13 @@ exports.deleteTask = function(request, response) {
                 console.log("Error:", error);
             });
 
-            const promises = [p1, p2, p3, p4];
+            const sendMailToWatcherPromise = getTask(taskId, orgDomain).then((taskDoc) => {
+                taskDoc.Watcher.forEach(element => {
+                    watchers.push(element);
+                });
+            });
+
```
Comment (viv9k) [inline functions/model/tasks/tark/deleteTask.js:127]: No need for these changes, Also the logic can be improved. For the time being please remove it.

File: functions/model/tasks/tark/deleteTask.js [hunk 3]
```diff
@@ -129,6 +135,10 @@ exports.deleteTask = function(request, response) {
...
 
+                watchers.forEach(element => {
+                    sendMail(element, subjectMessage, htmlMessage);
+                    console.log("mail sent!!!!!!!")
+                });
```
Comment (viv9k) [inline functions/model/tasks/tark/deleteTask.js:141]: Same here

File: functions/model/tasks/tark/editTask.js [hunk 4]
```diff
@@ -223,13 +227,23 @@ exports.editTask = function(request, response) {
...
         });
 
-        promises.push(p3);
+        const sendMailToWatcherPromise = getTask(taskId, orgDomain).then((taskDoc) => {
+            taskDoc.Watcher.forEach(element => {
+                watchers.push(element);
+            });
+        });
+
+        promises.push(p3, sendMailToWatcherPromise);
```
Comment (viv9k) [inline functions/model/tasks/tark/editTask.js:236]: Same here

File: functions/model/tasks/tark/editTask.js [hunk 5]
```diff
@@ -126,6 +127,9 @@ exports.editTask = function(request, response) {
             const p2 = getTask(taskId, orgDomain).then((taskDoc) => {
                 const project = taskDoc.Project;
                 const teamId = taskDoc.TeamId;
+                taskDoc.Watcher.forEach(element => {
+                    watchers.push(element);
+                });
```
Comment (viv9k) [inline functions/model/tasks/tark/editTask.js:132]: Same here

File: functions/model/tasks/tark/editTask.js [hunk 6]
```diff
@@ -223,13 +227,23 @@ exports.editTask = function(request, response) {
...
         updateTask(updateTaskJson, orgDomain, taskId);
+        watchers.forEach(element => {
+            sendMail(element, subjectMessage, htmlMessage);
+            console.log("mail sent!!!!!!!")
+        });
```
Comment (viv9k) [inline functions/model/tasks/tark/editTask.js:246]: Same here

File: functions/model/tasks/tark/logwork.js [hunk 7]
```diff
@@ -76,12 +77,22 @@ exports.logWork = function(request, response) {
                 CompletionDate: completiondate,
             };
             updateTask(logWorkInputJson, orgDomain, taskId);
+            watchers.forEach(element => {
+                sendMail(element, subjectMessage, htmlMessage);
+                console.log("mail sent!!!!!!!")
+            });
```
Comment (viv9k) [inline functions/model/tasks/tark/logwork.js:83]: It's not gonna work, watcher is not filled

File: functions/model/tasks/tark/logwork.js [hunk 8]
```diff
@@ -76,12 +77,22 @@ exports.logWork = function(request, response) {
...
             console.log("Error:", error);
         });
-        promises.push(promise1);
+
+        const sendMailToWatcherPromise = getTask(taskId, orgDomain).then((taskDoc) => {
+            taskDoc.Watcher.forEach(element => {
+                watchers.push(element);
+            });
+        });
+        promises.push(promise1, sendMailToWatcherPromise);
```
Comment (viv9k) [inline functions/model/tasks/tark/logwork.js:95]: Same here

File: src/app/header/login-handler/user-card/user-card.component.ts [hunk 9]
```diff
@@ -25,10 +25,25 @@ import { StartServiceService } from 'src/app/services/start/start-service.servic
...
+  readProfileDetails(){
+    // this.photoUrl = this.authService.userAppSetting.photoURL
+    // const data = this.userService.getUserData(this.user.email);
+    // console.log("111",data)
+
+    // if(data != undefined) {
+    //   this.photoUrl = data.photoURL;
+    //   console.log("hereee:",this.photoUrl)
+    // }
+  }
```
Comment (viv9k) [inline src/app/header/login-handler/user-card/user-card.component.ts:44]: Unnecessary function

**Review by viv9k — CHANGES_REQUESTED** [review]
Please all the unnecessary changes

File: functions/model/tasks/tark/deleteTask.js (hunk 2 above)
Comment (printf-twinkle) [reply functions/model/tasks/tark/deleteTask.js:127]: removed

File: functions/model/tasks/tark/addComment.js (hunk 1 above)
Comment (printf-twinkle) [reply functions/model/tasks/tark/addComment.js:91]: done

File: functions/model/tasks/tark/deleteTask.js (hunk 3 above)
Comment (printf-twinkle) [reply functions/model/tasks/tark/deleteTask.js:141]: done

File: functions/model/tasks/tark/editTask.js (hunk 4 above)
Comment (printf-twinkle) [reply functions/model/tasks/tark/editTask.js:236]: done

File: functions/model/tasks/tark/editTask.js (hunk 5 above)
Comment (printf-twinkle) [reply functions/model/tasks/tark/editTask.js:132]: done

File: functions/model/tasks/tark/editTask.js (hunk 6 above)
Comment (printf-twinkle) [reply functions/model/tasks/tark/editTask.js:246]: done

File: functions/model/tasks/tark/logwork.js (hunk 7 above)
Comment (printf-twinkle) [reply functions/model/tasks/tark/logwork.js:83]: done

File: functions/model/tasks/tark/logwork.js (hunk 8 above)
Comment (printf-twinkle) [reply functions/model/tasks/tark/logwork.js:95]: done

File: src/app/header/login-handler/user-card/user-card.component.ts (hunk 9 above)
Comment (printf-twinkle) [reply src/app/header/login-handler/user-card/user-card.component.ts:44]: removed

File: functions/model/tasks/tark/addComment.js [hunk 10]
```diff
@@ -62,41 +63,40 @@ exports.addComment = function(request, response) {
...
 
-            addActivity("COMMENT", logWorkComment, taskId, date, time, orgDomain, uid);
+            addActivity("COMMENT", logWorkComment, taskId, date, time, orgDomain, uid);;
```
Comment (viv9k) [inline functions/model/tasks/tark/addComment.js:81]: 2 semicolons

File: functions/model/tasks/tark/addComment.js [hunk 11]
```diff
@@ -62,41 +63,40 @@ exports.addComment = function(request, response) {
...
             return error;
         });
-
         const updateTaskJson = {
             LastUpdatedDate: date,
         };
         updateTask(updateTaskJson, orgDomain, taskId);
 
-        const promises = [promise1, promise2];
+        const promises = [promise1, promise2, sendMailToWatcherPromise];
```
Comment (viv9k) [inline functions/model/tasks/tark/addComment.js:71]: sendMain is no more there

File: functions/model/tasks/tark/createTask.js [hunk 12]
```diff
@@ -28,7 +28,7 @@ const { getOrgUseAppKey } = require("../../organization/lib");
 const { getOrgRawData, updateOrgRawData } = require("../../organization/lib");
 const { getSprint, updateSprint, setSprint } = require("../../sprints/lib");
 const { updateTeamDetails, getTeam } = require("../../teams/lib");
-const { setTask } = require("../lib");
+const { setTask, getTask } = require("../lib");
```
Comment (viv9k) [inline functions/model/tasks/tark/createTask.js:31]: No use of getTask here

File: src/app/header/notification-center/notification-center.component.ts [hunk 13]
```diff
@@ -26,18 +28,56 @@ export class NotificationCenterComponent implements OnInit {
...
+      },
+      error: (error) => {
+        console.error("active notifications reset");
+      },
+      complete: () => console.info("notification reset successfully!")
+    })
+  }
+
+  showOlderNotifications() {
+    console.log("1")
```
Comment (viv9k) [inline src/app/header/notification-center/notification-center.component.ts:76]: remove unnecessary console

**Review by viv9k — CHANGES_REQUESTED** [review]
Please do the required changes

File: functions/model/tasks/tark/addComment.js (hunk 10 above)
Comment (printf-twinkle) [reply functions/model/tasks/tark/addComment.js:81]: done

File: functions/model/tasks/tark/addComment.js (hunk 11 above)
Comment (printf-twinkle) [reply functions/model/tasks/tark/addComment.js:71]: done

File: functions/model/tasks/tark/createTask.js (hunk 12 above)
Comment (printf-twinkle) [reply functions/model/tasks/tark/createTask.js:31]: done

File: src/app/header/notification-center/notification-center.component.ts (hunk 13 above)
Comment (printf-twinkle) [reply src/app/header/notification-center/notification-center.component.ts:76]: done

**Review by viv9k — APPROVED** [review]
Approved

**Pull request merged**